core/src/main/scala/scalacache/package.scala

```diff
@@ -50,21 +59,38 @@ package object scalacache extends JavaSerializationCodec {
           }
       }
 
-      val fromCache: Future[Option[From]] = getWithKey(key).recover[Option[From]] {
-        case e =>
-          if (logger.isWarnEnabled) {
-            logger.warn(s"Failed to read from cache. Key = $key", e)
-          }
-          None
+      def calculateAndCacheResult(): Future[From] = {
+        val result: Future[From] = f
+        asynchronouslyCacheResult(result)
+        result
       }
 
-      fromCache flatMap {
-        case Some(value) => Future.successful(value)
-        case None =>
-          val result: Future[From] = f
-          asynchronouslyCacheResult(result)
-          result
+      val fromCache: Future[Option[From]] = getWithKey(key)
+
+      if (fromCache.isCompleted) {
+        // optimisation for in-memory caches that return Future.successful(...)
+        fromCache.value.get match {
+          case Success(Some(value)) => Future.successful(value)
```

Comment:
It seems unfortunate that every read creates a wrapper, especially when some caches support memoization directly. Perhaps you could push the `calculateAndCacheResult` into the non-memoizing caches so that the returned future is bubbled directly to the caller?